File: backend/internal/service/deployment_types.go
package service

import "time"

type CreateDeploymentCommand struct {
	RequesterUserID string
	RequesterRole   string
	ProjectID       string
	BlueprintID     string
	TriggerKind     string
	SourceRef       string
	Artifact        BlueprintArtifactMetadata
	ServiceIDs      []string
}

type DesiredStateRevisionRecord struct {
	ID                   string
	ProjectID            string
	ProjectSlug          string
	BlueprintID          string
	DeploymentBindingID  string
	Namespace            string
	CommitSHA            string
	ArtifactRef          string
	ImageRef             string
	TriggerKind          string
	Status               string
	RuntimeMode          string
	Services             []BlueprintServiceContractRecord
	ServiceSpecs         []K3sServiceSpecRecord
	DependencyBindings   []LazyopsYAMLDependencyBinding
	InternalBindings     []InternalBindingRecord
	CompatibilityPolicy  LazyopsYAMLCompatibilityPolicy
	MagicDomainPolicy    LazyopsYAMLMagicDomainPolicy
	ScaleToZeroPolicy    LazyopsYAMLScaleToZeroPolicy
	ManifestBundle       K3sManifestBundleRecord
	PublicDomains        []PublicDomainRecord
	PlacementAssignments []PlacementAssignmentRecord
	CreatedAt            time.Time
	UpdatedAt            time.Time
}

type DeploymentRecord struct {
	ID          string
	ProjectID   string
	RevisionID  string
	Status      string
	StartedAt   *time.Time
	CompletedAt *time.Time
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type CreateDeploymentResult struct {
	Revision   DesiredStateRevisionRecord
	Deployment DeploymentRecord
}

type DeploymentTimelineEventRecord struct {
	Timestamp   time.Time
	State       string
	Label       string
	Description string
}

type DeploymentOverviewRecord struct {
	ID                   string
	ProjectID            string
	RevisionID           string
	Revision             int
	CommitSHA            string
	ArtifactRef          string
	ImageRef             string
	TriggerKind          string
	BuildState           string
	RolloutState         string
	Promoted             bool
	TriggeredBy          string
	Namespace            string
	RuntimeMode          string
	Services             []BlueprintServiceContractRecord
	ServiceSpecs         []K3sServiceSpecRecord
	PlacementAssignments []PlacementAssignmentRecord
	PublicURLs           []string
	PublicURLReason      string
	StartedAt            *time.Time
	CompletedAt          *time.Time
	CreatedAt            time.Time
	UpdatedAt            time.Time
}

type DeploymentDetailRecord struct {
	DeploymentOverviewRecord
	Timeline        []DeploymentTimelineEventRecord
	CanRollback     bool
	CanPromote      bool
	CanCancel       bool
	SafetyPolicy    DeploymentSafetyPolicyRecord
	IncidentSummary *DeploymentIncidentSummaryRecord
}

type DeploymentSafetyPolicyRecord struct {
	AutoRollbackEnabled bool     `json:"auto_rollback_enabled"`
	Triggers            []string `json:"triggers"`
	Description         string   `json:"description"`
}

type DeploymentFixActionRecord struct {
	ID     string `json:"id"`
	Label  string `json:"label"`
	Href   string `json:"href"`
	Method string `json:"method"`
}

type DeploymentIncidentSummaryRecord struct {
	State         string                     `json:"state"`
	Headline      string                     `json:"headline"`
	Reason        string                     `json:"reason"`
	Recommended   string                     `json:"recommended"`
	IncidentID    string                     `json:"incident_id,omitempty"`
	IncidentKind  string                     `json:"incident_kind,omitempty"`
	IncidentLevel string                     `json:"incident_level,omitempty"`
	PrimaryAction *DeploymentFixActionRecord `json:"primary_action,omitempty"`
}

File: frontend/src/modules/deployments/deployment-types.ts
export type BuildState = 'draft' | 'queued' | 'building' | 'artifact_ready' | 'planned' | 'applying' | 'promoted' | 'failed' | 'rolled_back' | 'superseded';
export type RolloutState = 'queued' | 'running' | 'candidate_ready' | 'promoted' | 'failed' | 'rolled_back' | 'canceled';

export type DeploymentService = {
  name: string;
  path: string;
  public: boolean;
  runtime_profile: string;
};

export type PlacementAssignment = {
  service_name: string;
  target_id: string;
  target_kind: string;
  labels: Record<string, string>;
};

export type DeploymentRecord = {
  id: string;
  project_id: string;
  revision_id: string;
  revision: number;
  commit_sha: string;
  artifact_ref: string | null;
  image_ref: string | null;
  trigger_kind: string;
  build_state: BuildState;
  rollout_state: RolloutState;
  promoted: boolean;
  triggered_by: string;
  runtime_mode: string;
  services: DeploymentService[];
  placement_assignments: PlacementAssignment[];
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type DeploymentListResponse = {
  items: DeploymentRecord[];
};

export type DeploymentTimelineEvent = {
  timestamp: string;
  state: string;
  label: string;
  description: string;
};

export type DeploymentSafetyPolicy = {
  auto_rollback_enabled: boolean;
  triggers: string[];
  description: string;
};

export type DeploymentFixAction = {
  id: string;
  label: string;
  href: string;
  method: string;
};

export type DeploymentIncidentSummary = {
  state: string;
  headline: string;
  reason: string;
  recommended: string;
  incident_id?: string;
  incident_kind?: string;
  incident_level?: string;
  primary_action?: DeploymentFixAction;
};

export type DeploymentDetail = DeploymentRecord & {
  timeline: DeploymentTimelineEvent[];
  can_rollback: boolean;
  can_promote: boolean;
  can_cancel: boolean;
  safety_policy: DeploymentSafetyPolicy;
  incident_summary?: DeploymentIncidentSummary;
};
